servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ValidateDemandPublisher.java

```diff
@@ -0,0 +1,107 @@
+    private final LongBinaryConsumer requestConsumer;
+
+    ValidateDemandPublisher(final Publisher<T> original,
+                            final ObjLongConsumer<T> onNextConsumer,
+                            final LongBinaryConsumer requestConsumer) {
+        super(original);
+        this.onNextConsumer = requireNonNull(onNextConsumer);
+        this.requestConsumer = requireNonNull(requestConsumer);
+    }
+
+    @Override
+    public Subscriber<? super T> apply(final Subscriber<? super T> subscriber) {
+        return new ValidateDemandSubscriber<>(this, subscriber);
+    }
+
+    private static final class ValidateDemandSubscriber<T> implements Subscriber<T> {
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<ValidateDemandSubscriber> demandUpdater =
+                AtomicLongFieldUpdater.newUpdater(ValidateDemandSubscriber.class, "demand");
+        private final Subscriber<? super T> subscriber;
+        private final ValidateDemandPublisher<T> parent;
+        private volatile long demand;
+
+        private ValidateDemandSubscriber(ValidateDemandPublisher<T> parent,
+                                         final Subscriber<? super T> subscriber) {
+            this.parent = parent;
+            this.subscriber = requireNonNull(subscriber);
+        }
+
+        @Override
+        public void onSubscribe(final Subscription subscription) {
+            subscriber.onSubscribe(new Subscription() {
+                @Override
+                public void request(final long n) {
+                    try {
+                        if (isRequestNValid(n)) {
+                            final long currDemand = demandUpdater.accumulateAndGet(ValidateDemandSubscriber.this, n,
+                                    FlowControlUtils::addWithOverflowProtection);
+                            parent.requestConsumer.accept(n, currDemand);
+                        }
+                    } finally {
+                        subscription.request(n);
+                    }
+                }
+
+                @Override
+                public void cancel() {
+                    subscription.cancel();
+                }
+            });
+        }
+
+        @Override
+        public void onNext(@Nullable final T t) {
+            final long currDemand = demandUpdater.decrementAndGet(this);
+            parent.onNextConsumer.accept(t, currDemand);
+            if (currDemand < 0) {
+                throw new InvalidDemandException("Received onNext signal='" + t + "' with no demand");
+            }
+            subscriber.onNext(t);
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            subscriber.onError(t);
+        }
+
+        @Override
+        public void onComplete() {
+            subscriber.onComplete();
+        }
+    }
+
+    static final class InvalidDemandException extends IllegalStateException {
+        InvalidDemandException(String message) {
```

Comment:
Our users found it hard to do anything with custom exceptions that are not public. In this case, they will see `InvalidDemandException` in logs, but won't have any clue what to do with it until they look at the source code and see that it's actually an `IllegalStateException`. Consider throwing `IllegalStateException` exception instead or make this one public.

Reply:
`InvalidDemandException` helps validate the test condition but I will go with `IllegalStateException` for now if its an issue.